src/log.h

```diff
@@ -221,16 +279,62 @@ class StreamProxy {
 public:
 	StreamProxy(std::ostream *os) : m_os(os) { }
 
+	inline static void maybe_fix_stream_state(std::ostream &os)
+	{
+		std::ios::iostate state = os.rdstate();
+		if (state != std::ios::goodbit) {
+			// clear error state so the stream works again
+			os.clear();
+			if (state & std::ios::eofbit)
+				os << "(ostream:eofbit)";
+			if (state & std::ios::badbit)
+				os << "(ostream:badbit)";
+			if (state & std::ios::failbit)
+				os << "(ostream:failbit)";
+		}
```

Comment:
this would be good enough IMO
```suggestion
		if (!os.good()) {
			// clear error state so the stream works again
			os.clear();
			os << "(ostream error)";
		}
```

Reply:
I strongly disagree. If a future user or dev sees `(ostream error)` in the log. They will have no idea what kind of error that means. If they google this, there are no helpful results, because it is too generic. They will have to grep the codebase and then read through `log.h` to figure out what it means. And then potentially add more code to figure out exactly which error bit was set, and re-run minetest to generate a new log which contains this detail.

But if it says "(ostream:failbit)", and they google that or even just `failbit`, they will get the top result on cppreference, and immediately have a good idea of what happened. There's no need to go searching the logging code for clues.

Reply:
> They will have to grep the codebase

I don't think that's unnatural at all. But your choice.